src/main/java/uk/ac/cam/cl/dtg/isaac/api/services/EmailService.java

```diff
@@ -0,0 +1,130 @@
+
+public class EmailService {
+
+    public interface AssignmentLike {
+        Long getGroupId();
+
+        Long getOwnerUserId();
+
+        @Nullable
+        Date getDueDate();
+    }
+
+    public interface HasTitleOrId {
+        String getId();
+        @Nullable String getTitle();
+    }
+
+    private static final Logger log = LoggerFactory.getLogger(EmailService.class);
+
+    private final EmailManager emailManager;
+    private final GroupManager groupManager;
+    private final UserAccountManager userManager;
+
+    @Inject
+    public EmailService(final EmailManager emailManager, final GroupManager groupManager, final UserAccountManager userManager) {
+        this.emailManager = emailManager;
+        this.groupManager = groupManager;
+        this.userManager = userManager;
+    }
+
+    private static final DateFormat DATE_FORMAT = new SimpleDateFormat("dd/MM/yy");
+    public void sendAssignmentEmailToGroup(final AssignmentLike assignment, final HasTitleOrId on, final Map<String, Object> tokenToValueMapping, final String templateName) throws SegueDatabaseException {
+        UserGroupDTO userGroupDTO = groupManager.getGroupById(assignment.getGroupId());
+
+        String dueDate = "";
+        if (assignment.getDueDate() != null) {
+            dueDate = String.format(" (due on %s)", DATE_FORMAT.format(assignment.getDueDate()));
+        }
+
+        String name = on.getId();
+        if (on.getTitle() != null) {
+            name = on.getTitle();
+        }
+
+        RegisteredUserDTO assignmentOwnerDTO = null;
+        try {
+            assignmentOwnerDTO = this.userManager.getUserDTOById(assignment.getOwnerUserId());
+            String groupName = getFilteredGroupNameFromGroup(userGroupDTO);
+            String assignmentOwner = getTeacherNameFromUser(assignmentOwnerDTO);
+
+            final Map<String, Object> map = ImmutableMap.<String, Object>builder()
+                .put("gameboardName", name) // Legacy name
+                .put("assignmentName", name)
+                .put("assignmentDueDate", dueDate)
+                .put("groupName", groupName)
+                .put("assignmentOwner", assignmentOwner)
+                .putAll(tokenToValueMapping)
+                .build();
+
+            sendTemplatedEmailToActiveGroupMembers(userGroupDTO, templateName, map, EmailType.ASSIGNMENTS);
+        } catch (NoUserException e) {
+            log.error("Could not send assignment email because owner did not exist.", e);
+        }
+    }
+
+    public void sendTemplatedEmailToActiveGroupMembers(final UserGroupDTO userGroupDTO, final String templateName, final Map<String, Object> tokenToValueMapping, final EmailType emailType) throws SegueDatabaseException {
+        List<RegisteredUserDTO> usersToEmail = Lists.newArrayList();
+        Map<Long, GroupMembershipDTO> userMembershipMapforGroup = this.groupManager.getUserMembershipMapForGroup(userGroupDTO.getId());
```

Comment:
Am I right in thinking that this won't be the second time the group members are loaded from the database, but replaces the need to load them in the Assignments Manager? I got confused and thought it would be an additional load of all the users, but now I review it again I think it's ok?

Reply:
Yes, the idea is not to load them in `AssignmentManager` but to have this piece of code worry about the specifics of emailing a group.